core/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import io.grpc.ExperimentalApi;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.Socket;
+import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.security.spec.InvalidKeySpecException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.X509ExtendedKeyManager;
+
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/8024")
+/**
+ * AdvancedTlsX509KeyManager is an {@code X509ExtendedKeyManager} that allows users to configure
+ * advanced TLS features, such as private key and certificate chain reloading, etc.
+ */
+public final class AdvancedTlsX509KeyManager extends X509ExtendedKeyManager {
+  private static final Logger log = Logger.getLogger(AdvancedTlsX509KeyManager.class.getName());
+
+  // The private key and the cert chain we will use to send to peers to prove our identity.
+  private volatile PrivateKey key;
+  private volatile X509Certificate[] certs;
+
+  public AdvancedTlsX509KeyManager() { }
+
+  @Override
+  public PrivateKey getPrivateKey(String alias) {
+    // Same question as the trust manager: is it really thread-safe to do this?
+    return this.key;
+  }
+
+  @Override
+  public X509Certificate[] getCertificateChain(String alias) {
+    return this.certs;
+  }
+
+  @Override
+  public String[] getClientAliases(String keyType, Principal[] issuers) {
+    return new String[0];
+  }
+
+  @Override
+  public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
+    return "default";
+  }
+
+  @Override
+  public String chooseEngineClientAlias(String[] keyType, Principal[] issuers, SSLEngine engine) {
+    return "default";
+  }
+
+  @Override
+  public String[] getServerAliases(String keyType, Principal[] issuers) {
+    return new String[0];
+  }
+
+  @Override
+  public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+    return "default";
+  }
+
+  @Override
+  public String chooseEngineServerAlias(String keyType, Principal[] issuers,
+      SSLEngine engine) {
+    return "default";
+  }
+
+  /**
+   * Updates the current cached private key and cert chains.
+   * Right now, callers should make sure |key| matches the public key on the leaf certificate of
+   * |certs|.
+   * TODO(ZhenLian): explore possibilities to do a crypto check here.
```

Comment:
Move this out of the javadoc? The "right now" and TODO make it hard to use for the moment. If we add verification in the future that will mean this may start throwing, yet it isn't documented to throw at all today. How soon might that verification happen?

Reply:
Sure, I moved those sentences out of the javadoc.
The verification won't take too much time - it should just behave like a normal function call. The way it does this is to extract some particular values from both the private key and certificate, and then compare them to see if they are equal(See https://stackoverflow.com/a/24167548).